src/main/scala/rules/Evaluator.scala

```diff
@@ -943,38 +955,55 @@ object evaluator extends EvaluationRules {
                              assertReadAccessOnly = if (Verifier.config.respectFunctionPrePermAmounts())
                                s2.assertReadAccessOnly /* should currently always be false */ else true)
             consumes(s3, pres, true, _ => pvePre, v2)((s4, snap, v3) => {
-              val snap1 = snap.get.convert(sorts.Snap)
-              val preFApp = App(functionSupporter.preconditionVersion(v3.symbolConverter.toFunction(func)), snap1 :: tArgs)
-              val preExp = Option.when(withExp)({
-                DebugExp.createInstance(Some(s"precondition of ${func.name}(${eArgsNew.get.mkString(", ")}) holds"), None, None, InsertionOrderedSet.empty)
-              })
-              v3.decider.assume(preFApp, preExp)
+              val snap1 = v.decider.assumeSortWrapper(snap.get.convert(sorts.Snap))
               val funcAnn = func.info.getUniqueInfo[AnnotationInfo]
-              val tFApp = funcAnn match {
-                case Some(a) if a.values.contains("opaque") =>
-                  val funcAppAnn = fapp.info.getUniqueInfo[AnnotationInfo]
-                  funcAppAnn match {
-                    case Some(a) if a.values.contains("reveal") => App(v3.symbolConverter.toFunction(func), snap1 :: tArgs)
-                    case _ => App(functionSupporter.limitedVersion(v3.symbolConverter.toFunction(func)), snap1 :: tArgs)
-                  }
-                case _ => App(v3.symbolConverter.toFunction(func), snap1 :: tArgs)
+              def getFApp(fun: Function, args: Seq[Term]): Term = {
+                funcAnn match {
+                  case Some(a) if a.values.contains("opaque") =>
+                    val funcAppAnn = fapp.info.getUniqueInfo[AnnotationInfo]
+                    funcAppAnn match {
+                      case Some(a) if a.values.contains("reveal") => App(fun, args)
+                      case _ => App(fun, args )
+                    }
+                  case _ => App(fun, args)
+                }
               }
+
+              val fun = v3.symbolConverter.toFunction(func);
+              val res = if(s3.functionData(func).phase == 0) {
+                val r = v3.decider.fresh(func.result)
+                v3.decider.assume(Equals(r._1, getFApp(functionSupporter.limitedVersion(fun), snap1 :: tArgs)), None)
```

Comment:
Why the fresh term here? Wouldn't this be equivalent to directly returning the func app with the limited version?